_posts/2020-03-06-ok-mauro-but-i-want-to-do-paging-and-sorting.md

```diff
@@ -0,0 +1,42 @@
+
+As you probably noticed, the approaches described to page data and sort data work in isolation, but immediately fall apart if we need to use them together. If the need is to apply "paging and sorting" in the same request, then the proposed solutions cannot be applied. Or, if implemented, can only result in a very chatty and performance-avert design.
+
+## Paging and sorting is misleading
+
+Data sorting needs to happen before paging can be applied. If we page and then sort, we're going to sort only the current page, which means that if the user transitions to another page, the sort doesn't make sense anymore. It should be "sorting and paging," but that sounds bad.
+
+The problem is that if we have to sort first and the request coming in is like `/products?orderBy=price,description` then, as we already said, sorting is not so easy. The request reveals a relationship between the service owning prices and the service owning descriptions. If data are not in the same storage, the only option if first to load all the prices and sort them, then handover the result set to the service that can load all the descriptions and sort the result set based on price and description. Only at this point paging can be applied.
+
+## That's a join, Mauro
+
+If we were to apply the outlined logic manually, it's like we're reinventing the wheel. And we've already discussed something very similar, lately.
+
+When we talked about searching in distributed systems, the suggestion was to ship data into a commonplace best suited to perform a search. The same rule applies here. If the need is to sort data across multiple services and apply paging to the sorted result set, it's way simpler to perform all these steps in a single place, for example, a SQL database or an Elasticsearch instance.
+
+[Tomek](https://twitter.com/masternak) and I discussed at length how to do that in [The Quest for better Search](https://milestone.topics.it/soa-search/2019/05/15/the-quest-for-better-search.html) and the follow-up [Search is a Team Effort](https://milestone.topics.it/soa-search/2019/05/22/search-is-a-team-effort.html). It's just a matter of applying the same logic in this case.
+
+## Conclusion
+
+When designing distributed systems, one of the goals is to keep coupling as low as possible. Zero coupling is an illusion, and we want to have in our tool belt all the tools that we might need to keep coupling low. Do we need to apply paging? We understood how to do that with near to zero coupling. Do we need to sort data? It can be done. Do we need to sort and page data? Sure, there is a solution to that as well, but this time comes with a little coupling. As for searching, we want to be very careful and make sure that only data required to apply the sorting end up being stored in the storage. We don't want to pay the price of unneeded coupling.
```

Comment:
```suggestion
When designing distributed systems, one of the goals is to keep coupling as low as possible. Zero coupling is an illusion, and we want to have in our tool belt all the tools that we might need to keep coupling low. Do we need to apply paging? We understood how to do that with almost no coupling. Do we need to sort data? It can be done. Do we need to sort and page data? Sure, there is a solution to that as well, but this time comes with a little coupling. As for searching, we want to be very careful and make sure that only data required to apply the sorting end up being stored in the storage. We don't want to pay the price of unneeded coupling.
```